src/libraries/System.Private.CoreLib/src/System/DateTime.cs

```diff
@@ -1408,7 +1409,7 @@ public static bool TryParseExact(ReadOnlySpan<char> s, ReadOnlySpan<char> format
             return DateTimeParse.TryParseExact(s, format, DateTimeFormatInfo.GetInstance(provider), style, out result);
         }
 
-        public static bool TryParseExact(string? s, string?[]? formats, IFormatProvider? provider, DateTimeStyles style, out DateTime result)
+        public static bool TryParseExact([NotNullWhen(true)] string? s, string?[]? formats, IFormatProvider? provider, DateTimeStyles style, out DateTime result)
```

Comment:
I'm not sure it really matters for any meaningful usage, but [NotNullWhen(true)] could also be on `formats` I believe.

Reply:
I thought `NotNullWhen` describes outputs based on the return value. How does that make sense on the `s` parameter?

Reply:
Never mind, I just read the issue description. Seems odd, but I can buy that. Question is whether the compiler understands that in all scenarios. /cc @jaredpar

Reply:
We've used it in limited capacity for this purpose, e.g. on string.IsNullOrEmpty (though that one is `[NotNullWhen(false)]` but same idea).  The question here is whether we should be embracing it more broadly in cases where it's not the purpose of the API.

Reply:
Not quite sure what you mean by "all scenarios" but this is an understood behavior. The most common use case we've found for this in Roslyn is extension methods. Pretty much all of our extension methods that only act on non-null *this* get marked with the simpler `[NotNull]`

Reply:
> Pretty much all of our extension methods that only act on non-null this get marked with the simpler [NotNull]

Why wouldn't they just be non-nullable? Are the cases you're referring to unconstrained generics?

Reply:
Parameters in general or specifically extension method receivers? 

Generally the compiler only moves a value to non-null when you take an action that would enforce this at runtime. So invoking a member for example. Passing a parameter is ambiguous. It's possible the method implementation did nothing with the parameter and hence the null state doesn't change. That's the basic summary of the logic here.

Reply:
If I have `public static void Foo(this object o);` with nullable enabled, and I do `object o = null; o.Foo();`, the compiler will warn... why is [NotNull] needed here?

Reply:
It's not needed but it enables other scenarios. Consider an expansion of your sample:

```cs
object? o = null;
o.Foo();
o.Foo(); // to warn or not to warn
```

If you use `[NotNull]` the second `o.Foo()` will not issue a warning. Example: 

https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYAYAEMEDoAiYAhgOYB2A9gM7BgDGVuAwhQCYCmAgmURAJ5UwVANypUAYjIBXCBCIAjCO0zsei9mLQBmbAgBs2OJgCiAD2CrBFMqgDeqTI+w6cBmABZMAWQQAKYAAWQpgU8gBW7HTAIQCUmLaYAL4OTjAu+tieXnC+ANoAchTA+TIQALqYgcGhEVGx8Ukpjk3Oum6eZkQAtgAOSr4xLfYoTqMh4ZHAAPwhmAC8mNKyoiNjjhS4PgPCmAD0u5gA7kQATjaraxtbMTv7R6fna80XYxTzi6UrTyGbOTctoyufy+TxayRQiSAA===

Reply:
Thanks.  It seems wrong to me that [DisallowNull] or [NotNull] should have any effect at all on a non-nullable parameter.

Reply:
You could certainly make an argument that we shouldn't ever double warn here. Essentially

> You gave me a warning that the value was null when I passed it to a parameter. That was enough. Just flip the state to non-null after. 

The counter is effectively what I said above

> Passing to a parameter doesn't ensure the value is non-null. We only flip the state to non-null when we are extremely certain the value is non-null and this doesn't meet the bar. 

That was definitely debated in LDM for nullable reference types ... extensively. The decision though was to not flip the state when it was passed as a parameter.

Reply:
My argument is [NotNull] should have zero impact here, as the parameter is already not-null.